### PR TITLE
Add analyser to detect cyclic relations

### PR DESCRIPTION
**Pull request by Famlam**
New analyser which detects self-referring relations, either directly (`r1 > r2 > r1`) or via other relations (`r1 > r2 > r3 > r1`).
Restricted to 10 referrals (but can be increased if needed) to avoid potential infinite loops, and limited to relations which have a location (otherwise you'll just get a NULL location).

It doesn't return many results (as self-referring relations are pretty rare; JOSM also doesn't allow adding them) but in nearly all cases it finishes within a second.
[Some test results attached](https://github.com/osm-fr/osmose-backend/files/10254834/cyclic.zip)

#1635

### CODE REVIEW

**Comment by jocelynj** [comment]
That's a very nice idea! And I like the fact that a new test was added :)

File: analysers/analyser_osmosis_relation_cyclic.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  SELECT
+    rcte.start_id,
+    child.member_id,
+    nloops + 1 AS nloops,
+    CONCAT(rcte.path, ' > r', child.relation_id) AS path
+  FROM
+    rcte
+    JOIN relation_members AS child ON
+      child.relation_id = rcte.member_id
+    WHERE
```
Comment (frodrigo) [inline analysers/analyser_osmosis_relation_cyclic.py:52]:
```suggestion
      child.relation_id = rcte.member_id AND
```

File: analysers/analyser_osmosis_relation_cyclic.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+)
+SELECT DISTINCT ON (start_id)
+  start_id,
+  ST_AsText(relation_locate(start_id)),
+  path
+FROM
+  rcte
+WHERE
+  start_id = member_id AND
+  nloops < {maxloops}
```
Comment (frodrigo) [inline analysers/analyser_osmosis_relation_cyclic.py:65]: Always true, doesn't ?

File: analysers/analyser_osmosis_relation_cyclic.py (hunk 1 above)
Comment (Famlam) [reply analysers/analyser_osmosis_relation_cyclic.py:52]: Just curiosity, it seems either produces the same result (unlike `LEFT JOIN`s), is there a benefit to either of them? I picked this one as it seemed easier to read to me, but will adjust it. Just curious

File: analysers/analyser_osmosis_relation_cyclic.py (hunk 2 above)
Comment (Famlam) [reply analysers/analyser_osmosis_relation_cyclic.py]: Indeed, no need to filter for nloops given the first condition

**Comment by Famlam** [comment]
Not sure what happened with the "Merge branch '...' thing...

File: analysers/analyser_osmosis_relation_cyclic.py (hunk 1 above)
Comment (frodrigo) [reply analysers/analyser_osmosis_relation_cyclic.py]: ON and WHERE are very close, but not the same. Here it just for code readability. I personally prefer ON for condition on joined table, keep condition and table declaration close.

**Comment by frodrigo** [comment]
You need to add the new analyser in the config file.

File: analysers/analyser_osmosis_relation_cyclic.py (hunk 1 above)
Comment (Famlam) [reply analysers/analyser_osmosis_relation_cyclic.py]: Thanks!

**Comment by Famlam** [comment]
> You need to add the new analyser in the config file.

Thanks, done!
(I always run it with `--analyser=...`, hence I didn't notice)

**Comment by Famlam** [comment]
Interesting, apparently there are even quite a lot of self-referring relations, for example: `r2200182 > r2200182`
Anyway, results look ok :)

**Pull request merged**